### ISSUE TITLE
How to generate IR for systemc by using this tool?

### ISSUE THREAD

**Issue by Elwin-zhao** [anikau31/systemc-clang#402]
# Description

# Reproducing the bug 

## Versions 
- systemc-clang: [i.e. 0.1]
- CXX compiler: [i.e. clang-7.0.0]
- CXX flags: [i.e. llvm-config --cxxflags]
- SystemC: [i.e. 2.3.3]
- Operating System: [i.e. Ubuntu 18.04]

## Minimal Working Example (MWE)
Provide a minimal working example that illustrates the bug.
```c++
int main() { 
  // Code here.
  return 0;
}
```

## Steps to Reproduce Bug
1. Do this
2. Do that
3. Compile that

### Actual behaviour
Explain the actual behviour observed.

**Comment by rseac** [comment]
@Elwin-zhao  Have you been able to successfully install the toolchain?  Can you tell me how far you got through the generation?  

We have documentation here: https://systemc-clang.readthedocs.io/en/latest/ if you haven't seen it already.

**Comment by Elwin-zhao** [comment]
I have been able to successfully install the toolchain, and I have been able to use scripts/run.sh to generate symbol table of a  systemc program, but how to generate IR?

**Comment by rseac** [comment]
You can use the `scripts/run-cpp-sv.sh` script. You will see two files being generated. One of those will have `_hdl.txt`, which contains the HCode. Note that this script will generate both the SystemVerilog and the HCode.  The files will be placed at the same location where the source file exists.

**Comment by Elwin-zhao** [comment]
@rseac There is an error which is thrown after I use the scripts/run-cpp-sv.sh 
"raise SystemCClangFatalError('Target file: {} is not generated by systemc-clang binary'.format(t))"

**Comment by Elwin-zhao** [comment]
@rseac I'm afraid that invoking systemc-clang directly will not generate _hdl.txt. 
![截屏2023-09-01 16 46 38](https://github.com/anikau31/systemc-clang/assets/50231687/ac097feb-e116-44fd-ba62-448adc276997)
And I also try to use scripts/run-cpp-sv.sh to generate file, however there is an error:
![截屏2023-09-01 16 50 57](https://github.com/anikau31/systemc-clang/assets/50231687/03434d80-3e3a-4f9f-9322-e3738c1e28ea)

**Comment by Elwin-zhao** [comment]
@rseac And I have another question, after I generate IR, how can I turn it into executable binary file? As you know, we can write IR and use llvm to compile and execute it.
 **e.g. llvm-as test.ll -o output.bc, lli output.bc.** 
So after I generate IR by using systemc-clang, how I can compile and execute it?

**Comment by rseac** [comment]
> @rseac I'm afraid that invoking systemc-clang directly will not generate _hdl.txt. ![截屏2023-09-01 16 46 38](https://user-images.githubusercontent.com/50231687/264972228-ac097feb-e116-44fd-ba62-448adc276997.png) And I also try to use scripts/run-cpp-sv.sh to generate file, however there is an error: ![截屏2023-09-01 16 50 57](https://user-images.githubusercontent.com/50231687/264973392-03434d80-3e3a-4f9f-9322-e3738c1e28ea.png)

The first screenshot actually shows systemc-clang running.  So that's good. I want to check that you're building systemc-clang with the `-DHDL` flag?  That enables the HDL plugin.  You would provide this when running `cmake`.


> @rseac And I have another question, after I generate IR, how can I turn it into executable binary file? As you know, we can write IR and use llvm to compile and execute it. **e.g. llvm-as test.ll -o output.bc, lli output.bc.** So after I generate IR by using systemc-clang, how I can compile and execute it?

The IR called HCode in this tool is a custom IR. We currently have a backend that targets SystemVerilog.  You would use the `-DHDL` to generate the SystemVerilog too.

**Comment by Elwin-zhao** [comment]
@rseac I have enabled DHDL and I use run.sh to generate _hdl.txt. Actually the file _hdl.txt can be created but it is empty. Also I use run-cpp-sv.sh to generate _hdl.txt.v, the file is also empty

**Comment by mayagokhale** [comment]
Did you do a cmake with -DHDL=ON? (and delete the CMakeCache.txt in the build directory first?

**Comment by Elwin-zhao** [comment]
Suppose that I want to generate 

> > @rseac I'm afraid that invoking systemc-clang directly will not generate _hdl.txt. ![截屏2023-09-01 16 46 38](https://user-images.githubusercontent.com/50231687/264972228-ac097feb-e116-44fd-ba62-448adc276997.png) And I also try to use scripts/run-cpp-sv.sh to generate file, however there is an error: ![截屏2023-09-01 16 50 57](https://user-images.githubusercontent.com/50231687/264973392-03434d80-3e3a-4f9f-9322-e3738c1e28ea.png)
> 
> The first screenshot actually shows systemc-clang running. So that's good. I want to check that you're building systemc-clang with the `-DHDL` flag? That enables the HDL plugin. You would provide this when running `cmake`.
> 
> > @rseac And I have another question, after I generate IR, how can I turn it into executable binary file? As you know, we can write IR and use llvm to compile and execute it. **e.g. llvm-as test.ll -o output.bc, lli output.bc.** So after I generate IR by using systemc-clang, how I can compile and execute it?
> 
> The IR called HCode in this tool is a custom IR. We currently have a backend that targets SystemVerilog. You would use the `-DHDL` to generate the SystemVerilog too.

Suppose I want to generate LLVM-based IR, how should I modify the source code? HCode does not have module information which is important for systemc language, because I want to analyse IR and do some optimization such as conflict detection and handling @rseac @mayagokhale

**Comment by rseac** [comment]
@Elwin-zhao Let us address some of these issues one at a time. 

First, let us see if you're compiling with the HDL plugin.

Looking at your first screen capture, I see that you do not have an output to the console that shows `HDL plugin`. This means that you haven't compiled with the `-DHDL=on` flag. You should see an output such as:

```bash
 $  ./systemc-clang                                
HDL plugin
....
```
Do you see that output to the console?  Let us confirm this before we proceed to addressing the other issues.

**Comment by Elwin-zhao** [comment]
![截屏2023-10-12 17 04 21](https://github.com/anikau31/systemc-clang/assets/50231687/e596e147-4e69-472e-a0e7-d286a1bfad19)
I have compiled with the -DHDL=on but there is an error @rseac

**Comment by rseac** [comment]
@Elwin-zhao It's good that the `HDL plugin` is showing up. This means that the plugin compiled.

Can you try the command that you see but just remove the `-debug` flag and let me know what happens?

**Comment by Elwin-zhao** [comment]
> @Elwin-zhao It's good that the `HDL plugin` is showing up. This means that the plugin compiled.
> 
> Can you try the command that you see but just remove the `-debug` flag and let me know what happens?
![截屏2023-10-13 11 27 23](https://github.com/anikau31/systemc-clang/assets/50231687/2ad33702-5f1c-4a7a-8731-bd6d2391d3c2) @rseac

**Comment by rseac** [comment]
@Elwin-zhao  Did you compile clang from source or are you using the pre-built binaries?  I want to try and setup the same build on my side to try and reproduce this error.

**Comment by Elwin-zhao** [comment]
> that you're using clang 15.0.0. Could you try with 15.0.6? I have noticed when bumping up versions of clang that there were some surprising issues that got fixed in the clang source code ...
I compiled clang from source.
@rseac

**Comment by rseac** [comment]
@Elwin-zhao For now, can you try using the pre-built binaries for 15.0.6 from https://github.com/llvm/llvm-project/releases/tag/llvmorg-15.0.6?

Hiren

**Comment by Elwin-zhao** [comment]
Another question. Suppose I want to get the AST generated by Clang, how can I get it? Should I modify the source code? If so, how should I modify it? @rseac

**Comment by Elwin-zhao** [comment]
<img width="983" alt="截屏2023-10-20 17 38 48" src="https://github.com/anikau31/systemc-clang/assets/50231687/086d9d7d-fc63-49a2-9155-74862548533e">
There is an error when I bumped Clang to 15.0.6. How to fix this bug? @rseac

**Comment by Elwin-zhao** [comment]
> @Elwin-zhao For now, can you try using the pre-built binaries for 15.0.6 from https://github.com/llvm/llvm-project/releases/tag/llvmorg-15.0.6?
> 
> Hiren

This problem has been fixed after I using clang 15.0.6 @rseac

**Comment by rseac** [comment]
Great to hear this @Elwin-zhao
